libbpf-tools/schedtimes.c

```diff
@@ -0,0 +1,358 @@
+        }
+        env.tid = pid;
+        break;
+    case 'c':
+        errno = 0;
+        pid = strtol(arg, NULL, 10);
+        if (errno || pid <= 0) {
+            fprintf(stderr, "Invalid PPID: %s\n", arg);
+            argp_usage(state);
+        }
+        env.ppid = pid;
+        break;
+    case ARGP_KEY_ARG:
+        errno = 0;
+        if (pos_args == 0) {
+            interval = strtol(arg, NULL, 10);
+            if (errno || interval <= 0) {
+                fprintf(stderr, "invalid interval (in s)\n");
+                argp_usage(state);
+            }
+        } else if(pos_args == 1) {
+            count = strtol(arg, NULL, 10);
+            if (errno || count <= 0){
+                fprintf(stderr, "invalid count\n");
+                argp_usage(state);
+            }
+        }
+        else {
+            fprintf(stderr, "Unrecognized positional argument: %s\n", arg);
+            argp_usage(state);
+        }
+        pos_args++;
+        break;
+    default:
+        return ARGP_ERR_UNKNOWN;
+    }
+    return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+    if (level == LIBBPF_DEBUG && !env.verbose)
+        return 0;
+    return vfprintf(stderr, format, args);
+}
+
+static void sig_int(int signo)
+{
+    exiting = 1;
+}
+
+static int sort_column(const void *obj1, const void *obj2)
+{
+    __u64 s1_val, s2_val;
+    struct sched_times_t *s1 = (struct sched_times_t*)obj1;
+    struct sched_times_t *s2 = (struct sched_times_t*)obj2;
+
+
+
+    if(sort_by == RUN){
+        s2_val = s2->run_time;
+        s1_val = s1->run_time;
+    }else if(sort_by == SLEEP){
+        s2_val = s2->sleep_time;
+        s1_val = s1->sleep_time;
+    }else if(sort_by == BLOCK){
+        s2_val = s2->block_time;
+        s1_val = s1->block_time;
+    }else if(sort_by == QUEUE){
+        s2_val = s2->queue_time;
+        s1_val = s1->queue_time;
+    }else{
+        //default by total
+        s2_val = s2->run_time + s2->sleep_time + s2->block_time + s2->queue_time;
+        s1_val = s1->run_time + s1->sleep_time + s1->block_time + s1->queue_time;
+    }
+
+    //workaround int overflow for u64
+    if(s2_val > s1_val){
+        return 1;
+    }else{
+        return s2_val < s1_val ? -1 : 0;
+    }
+}
+
+static int print_map(struct schedtimes_bpf *obj)
+{
+    __u32 lookup_key = -2, next_key;
+    int schedtimes_fd;
+    struct sched_times_t val;
+    time_t t;
+    struct tm *tm;
+    char ts[15];
+
+    static struct sched_times_t datas[OUTPUT_ROWS_LIMIT];
+    int i, err=0, rows=0;
+
+    schedtimes_fd = bpf_map__fd(obj->maps.sched_times);
+
+    time(&t);
+    tm = localtime(&t);
+    strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+    printf("%s\n", ts);
+    printf("%-16s %-7s %-10s %-10s %-10s %-10s %s\n",
+            "COMM", "PID", "RUN(us)", "SLEEP(us)", "BLOCK(us)", "QUEUE(us)", "TOTAL(us)");
+
+
+    while (!bpf_map_get_next_key(schedtimes_fd, &lookup_key, &next_key)){
```

Comment:
We should avoid using `bpf_map_get_next_key` which might have some issues if map delete happens in between. Please check batch_map_ops/dump_hash in map_helpers.c.

Reply:
changed,  the code is shortter, with overhead of extra keys  memory cost.
btw: there's  no other delete  in schedtimes

and I think maybe we should add another clear_hash helper ?